cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -80,35 +89,81 @@ public RemoteTaskExecutor(CConfiguration cConf, RemoteClientFactory remoteClient
    * @throws Exception returned by remote task if any
    */
   public byte[] runTask(RunnableTaskRequest runnableTaskRequest) throws Exception {
+    //initialize start time for collecting latency metric
+    long startTime = System.currentTimeMillis();
     ByteBuffer requestBody = encodeTaskRequest(runnableTaskRequest);
 
-    return Retries.callWithRetries(() -> {
-      try {
-        HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.POST, TASK_WORKER_URL)
-          .withBody(requestBody.duplicate());
-        if (compression) {
-          requestBuilder.addHeader(HttpHeaders.CONTENT_ENCODING, "gzip");
-          requestBuilder.addHeader(HttpHeaders.ACCEPT_ENCODING, "gzip, deflate");
-        }
-        HttpRequest httpRequest = requestBuilder.build();
+    try {
+      byte[] result = Retries.callWithRetries((attempts) -> {
+        try {
+          HttpRequest.Builder requestBuilder = remoteClient
+            .requestBuilder(HttpMethod.POST, TASK_WORKER_URL)
+            .withBody(requestBody.duplicate());
+          if (compression) {
+            requestBuilder.addHeader(HttpHeaders.CONTENT_ENCODING, "gzip");
+            requestBuilder.addHeader(HttpHeaders.ACCEPT_ENCODING, "gzip, deflate");
+          }
+          HttpRequest httpRequest = requestBuilder.build();
 
-        HttpResponse httpResponse = remoteClient.execute(httpRequest);
-        if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+          HttpResponse httpResponse = remoteClient.execute(httpRequest);
+          if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+            throw new RetryableException(
+              String.format("Received response code %s for %s", httpResponse.getResponseCode(),
+                            runnableTaskRequest.getClassName()));
+          }
+          if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+            BasicThrowable basicThrowable = GSON
+              .fromJson(new String(getResponseBody(httpResponse)), BasicThrowable.class);
+            throw RemoteExecutionException.fromBasicThrowable(basicThrowable);
+          }
+          //emit metrics with successful result
+          emitMetrics(startTime, true, runnableTaskRequest, attempts);
+          return getResponseBody(httpResponse);
```

Comment:
You should call `getResponseBody` first before emitting metrics, since a method like that can fail with exception.

```
byte[] result = getResponseBody(...);
emitMetrics(...);
return result;
```